Implement case mapping internals

iainireland opened the pull request:
<!--
Thank you for your pull request to ICU4X!

Reminder: try to use [Conventional Comments](https://conventionalcomments.org/) to make comments clearer.

Please see https://github.com/unicode-org/icu4x/blob/main/CONTRIBUTING.md for general
information on contributing to ICU4X.
-->

Fixes #1226, except for the `ucase_toFull(Upper|Lower|Title|Folding)` functions. Those functions require a context string to be passed in, and still need some design; since they aren't needed for my irregexp work, I'm splitting them off into a separate issue.

Other than those, this is a fairly direct re-implementation of ucase.cpp. Some implementation notes:
1. Following ucase_props_data.h, CaseMapping has three components: a code point trie that contains simple data covering most characters, an exception table for other characters, and an unfold map for case closure of strings.
2. I used the same data representation for the exception table as ICU4C: a packed array of variable-length entries, stored as an array of u16. The data is validated on construction. 
3. In ICU4C, the unfold map is stored (roughly speaking) as a sorted array of key/value pairs, where both the keys and the values are fixed-length UTF16 strings. Using the raw data directly in Rust turned out to be awkward because Rust is natively UTF8, so after a bit of benchmarking, I used a ZeroMap instead. (I actually benchmarked LiteMap vs a direct port vs ZeroMap: ZeroMap was the slowest, but LiteMap was the fastest, and they were all pretty close. I went with ZeroMap for memory reasons and in the hopes that it may be optimized better in the future now that we're actually using it.)
4. I put the provider code in the existing uprops provider, because it seems related, and because both providers need the ability to read code point tries.
5. Most of the tests are in the provider, where the test data is available. Some functions that are unlikely to be public in the final API are public here so that we can test them in another crate. (Of the currently implemented functions, I think only to_upper/to_lower/to_title are likely to be public in the final API.) Once we've written higher-level code that makes use of the public-for-testing functions, we can make them internal-only.
6. The SetAdder trait is a placeholder.
7. The data is from my in-flight PR for icuexportdata (https://github.com/unicode-org/icu/pull/1938).

sffc commented on provider/uprops/src/casemapping.rs:
```diff
@@ -0,0 +1,203 @@
+        let trie_data = &toml.ucase.code_point_trie;
+        let trie_type: TrieType =
+            TrieType::try_from(trie_data.trie_type_enum_val).map_err(DataError::new_resc_error)?;
+        let trie_header = CodePointTrieHeader {
+            high_start: trie_data.high_start,
+            shifted12_high_start: trie_data.shifted12_high_start,
+            index3_null_offset: trie_data.index3_null_offset,
+            data_null_offset: trie_data.data_null_offset,
+            null_value: trie_data.null_value,
+            trie_type,
+        };
```
Suggestion: It would be nice to make a utility function to convert from the TOML CPT Header to the ICU4X CPT Header.  The utility function could live in this crate for now (perhaps in a "helpers" module) and be shared between both this code and the uprops code.  I see that you already re-use `super::enumerated::EnumeratedPropertyCodePointTrie` in your serde definition, so this should be fairly easy to do.

iainireland replied:
I went ahead and implemented TryFrom<&SerializedCodePointTrie> for CodePointTrieHeader.

sffc commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1066 @@
+}
+
+impl<'data> CaseMappingUnfoldData<'data> {
+    // Unfold data is exported by ICU as an array of 16-bit values, representing a short
+    // header followed by a two-column key/value table. The header indicates:
+    // - The number of rows.
+    // - The number of UTF16 code units per row.
+    // - The number of UTF16 code units in the first (key) column.
+    //   (The number of code units in the value column can be derived from the above.)
```
Suggestion: I think this sort of decoding should be done in icuexportdata.  The TOML files we export should be "human readable" except for a small, specific set of exceptions, like CPTs.  If the data is really just a map from strings to strings, then the TOML should be a map of strings to strings.  Consumers of the TOML can readily massage the data into whatever format they like, including going back to an ICU4C-like structure.

If this is too much work, consider this suggestion optional.

iainireland replied:
I'm going to skip this one. Given that the other two pieces of data in this file are a CPT (opaque) and a packed array of variable length data structures (extremely opaque), I don't think there's a lot of value in making this particular field human readable. I expect that the data provider on the ICU4X side will be the only consumer of this TOML.

sffc commented on provider/uprops/src/uprops_serde.rs:
```diff
@@ -78,3 +78,28 @@ pub mod enumerated {
+
+    #[derive(serde::Deserialize)]
+    pub struct Level1 {
+        pub code_point_trie: super::enumerated::EnumeratedPropertyCodePointTrie,
```
Nit: Rename `EnumeratedPropertyCodePointTrie` to something more general.

iainireland replied:
`uprops_serde::enumerated::EnumeratedPropertyCodePointTrie` -> `uprops_serde::SerializedCodePointTrie`

sffc commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1066 @@
+        if self.has_slot(base_idx, CaseMappingExceptionSlot::FullMappings) {
+            let mapping_string = self
+                .full_mapping_string(base_idx, FullMapping::Fold)
+                .expect("No unpaired surrogates");
```
Issue: No panics in runtime code (unless you can prove they are unreachable).

I assume this is runtime code, but if this is data builder code, you should hide it behind a flag such as `provider_transform_internals` like we do elsewhere.

iainireland replied:
(This is runtime code.)

The design here is that `try_from_icu` is the only way to build a CaseMapping. In that function, we call the same helper function on every full mapping string and propagate errors. It is therefore impossible (short of writing broken serialized data by hand and deserializing it) for this to fail.

I will change the expect message here to something like "Data validated on construction".

sffc replied:
`try_from_icu` runs in the data transformer, but the class could also be constructed from serialized postcard data, correct? If so, the current policy is no panicking for code reachable by malformed/malicious data.

sffc replied:
Related: https://github.com/unicode-org/icu4x/issues/1183

sffc replied:
(this is still unresolved and I'm unsure how to move forward)

sffc commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1066 @@
+
+    const CLOSURE_MAX_LENGTH: u32 = 0xf;
+
+    fn try_from_icu(raw: &[u16]) -> Result<(Self, Vec<u16>), Error> {
```
Nit: Hide transformer-specific code behind the feature `provider_transform_internals`

iainireland commented:
@hsivonen Zibi thought this patch might be relevant to you.

sffc commented on experimental/casemapping/src/internals.rs:
```diff
@@ -723,29 +716,64 @@ impl<'data> CaseMapping<'data> {
             unfold,
-        })
+        };
+	result.validate()?;
```
Nit: You still have some tabs here.  Surprised `rustfmt` doesn't fix it for you.

sffc commented:
Also, please fix the merge conflict.

iainireland commented:
This has been sitting long enough waiting for review that I'm mostly done the next chunk of work, which restructures some of the internals, so I'm going to convert this to a draft and we can review the whole thing at once.

iainireland commented:
Okay, here's the reworked version. As we discussed in the deep dive, it includes internal support for various operations needed to implement case closure and titlecase, but doesn't expose them as part of the public API.

sffc commented on experimental/casemapping/src/casemapping.rs:
```diff
@@ -0,0 +1,149 @@
+
+impl CaseMapping {
+    /// A constructor which takes a [`ResourceProvider`] and creates a [`CaseMapping`].
+    pub fn new<P>(provider: &P) -> Result<CaseMapping, DataError>
```
Nit:

```suggestion
    pub fn try_new<P>(provider: &P) -> Result<CaseMapping, DataError>
```

sffc commented on experimental/casemapping/src/casemapping.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /// A constructor which takes a [`ResourceProvider`] and creates a [`CaseMapping`] for the given locale.
+    pub fn new_with_locale<P>(provider: &P, locale: &Locale) -> Result<CaseMapping, DataError>
```
Nit:

```suggestion
    pub fn try_new_with_locale<P>(provider: &P, locale: &Locale) -> Result<CaseMapping, DataError>
```

sffc commented on experimental/casemapping/src/casemapping.rs:
```diff
@@ -0,0 +1,149 @@
+
+impl CaseMapping {
+    /// A constructor which takes a [`ResourceProvider`] and creates a [`CaseMapping`].
+    pub fn new<P>(provider: &P) -> Result<CaseMapping, DataError>
```
Suggestion (optional): Consider changing the name to `try_new_default()` or `try_new_global()` or something similar to indicate the assumption that this method is making.  If you can't come up with a sufficiently good name then leave it as `try_new()`.

iainireland replied:
I was thinking of this along the same lines as `String`, although there it's `String::new` and String::with_capacity` (not `new_with_capacity`). I have a slight preference for leaving this as `try_new` and describing the default locale behaviour in the doc comment instead of trying to fit it in the name, but if there are other cases in ICU4X where we use something like "try_new_default" then consistency is probably enough of an argument to sway me.

sffc commented on experimental/casemapping/src/casemapping.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /// Returns the lowercase mapping of the given `char`.
+    /// This function only implements simple and common mappings. Full mappings,
+    /// which can map one `char` to a string, are not included.
```
Here and elsewhere:

Praise: Thank you for stating this caveat in the docs.

Request: Please link to the method you should use if you want the full behavior.

sffc commented on experimental/casemapping/src/error.rs:
```diff
@@ -0,0 +1,40 @@
+use icu_codepointtrie::error::Error as CodePointTrieError;
+
+/// A list of possible errors for the [`CaseMapping`](crate::CaseMapping) struct
+#[derive(Display, Debug, PartialEq)]
```
Request: make this `Copy` if possible.  #1396

iainireland replied:
DecodeUTF16Error and CodePointTrieError are not `Copy`, and the latter is not yet `Clone`. Easy enough to make both of our own errors `Clone`, at least.

sffc replied:
Discussion: https://github.com/unicode-org/icu4x/issues/1396#issuecomment-1035509876

sffc commented on experimental/casemapping/src/error.rs:
```diff
@@ -0,0 +1,40 @@
+
+impl Error {
+    /// Creates a new validation error with the given reason
+    pub fn invalid<T>(reason: &'static str) -> Result<T, Self> {
```
Nit:

```suggestion
    pub(crate) fn invalid<T>(reason: &'static str) -> Result<T, Self> {
```

sffc commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+use crate::error::Error;
+use crate::internals::{ClosureSet, DotType, MappingKind};
+
+// This represents case mapping exceptions that can't be represented as a delta applied to
+// the original code point. Similar to ICU4C, data is stored as a u16 array. The codepoint
+// trie in CaseMapping stores offsets into this array. The u16 at that index contains a
+// set of flags describing the subsequent data.
```
Discussion: I've said this before verbally, but I'll put it in writing. It really would be a *lot* more elegant with regard to code readability, ease of validation, and consistency across ICU4X if you would make this `VarZeroVec<ExceptionItemULE>` with something along the lines of

```rust
#[derive(VarULE)]
struct ExceptionItemULE {
    header: /* this can be a structured header type */,
    replacements: str, // you can put your UTF-8 in here and elimiate the side table
}
```

I'm quite confident you could cut the size of this file, and probably some of the sister files, at least in half if you did it this way.

The *only* downside I'm aware of is that that this incurs an additional "hop" since we need to look at the index table.  I'd be interested to see if this is a big enough penalty to show up in benchmarks.

Another model that might solve the problem would be to store code point replacements in the side table along with string replacements.  You could then have `ZeroVec<ExceptionItemULE>` where `ExceptionItemULE` is a fixed-width type containing the indices into the string.  You then do not need any other lookups.

sffc commented on experimental/casemapping/src/provider.rs:
```diff
@@ -0,0 +1,33 @@
+use crate::internals::CaseMappingInternals;
+use icu_provider::yoke;
+
+pub mod key {
+    //! Resource keys for [`icu_casemapping`](crate)
+    use icu_provider::{resource_key, ResourceKey};
+
+    /// Key for case mapping data
+    pub const CASE_MAPPING_V1: ResourceKey = resource_key!("props/casemap@1");
+}
```
Nit: You can and should remove the `key` module now that the key is in the ResourceMarker.

iainireland commented:
@markusicu Review ping (hoping for any domain-specific comments you might have about case mapping)

iainireland commented:
@echeran Have you had a chance to look at this yet?

echeran commented on experimental/casemapping/src/exceptions_builder.rs:
```diff
@@ -0,0 +1,205 @@
+use std::collections::HashMap;
+use zerovec::{VarZeroVec, ZeroVec};
+
+// CaseMappingExceptionsBuilder consumes the exceptions data produced by
```
suggestion: add a comment about usage pattern for motivating use case

IIRC, this is intended to be used prior to runtime, when the data is being transformed. That contrasts with `UnicodeSetBuilder`, which is intended as a runtime thing. It might be good to document, since most `*Builder`s I've seen follow the fluent builder pattern for runtime user APIs.

iainireland replied:
Will fix in future patch.

echeran commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1113 @@
+        let is_turkic = options.exclude_special_i;
+        match (c, is_turkic) {
+            // Turkic mappings
+            ('\u{49}', true) => '\u{131}', // 0049; T; 0131; # LATIN CAPITAL LETTER I
```
suggestion: have named constants for chars, too

Could these chars have named consatants like lines 623-628 (more readable)?

iainireland replied:
I mostly followed the ICU4C code here. I can take another pass and clean it up (in a future patch).

echeran commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1113 @@
+    }
+
+    // These constants are used for hardcoded locale-specific foldings.
+    const I_DOT: &'static str = "\u{69}\u{307}";
```
suggestion(optional): use 4 hex digits for BMP -- `\u{XXXX}`

Only because these are code points, I'm only used to seeing code points described as U+XXXX with 4 or more hex digits.

FWIW, C++ either allows 4 or 8 hex digits in Unicode char escapes: https://en.cppreference.com/w/cpp/language/escape

iainireland replied:
Will consider for future patch.

echeran commented on provider/uprops/src/casemapping.rs:
```diff
@@ -0,0 +1,271 @@
+
+    #[test]
+    fn test_simple_mappings() {
+        let root_dir = icu_testdata::paths::data_root()
+            .join("uprops")
```
[`uprops_toml_root()`](https://github.com/unicode-org/icu4x/blob/main/provider/testdata/src/paths.rs#L18)

iainireland replied:
Oh, cool, thanks!

echeran commented on provider/uprops/src/casemapping.rs:
```diff
@@ -0,0 +1,271 @@
+    // These tests are taken from StringCaseTest::TestCaseConversion in ICU4C.
+    #[test]
+    fn test_full_mappings() {
+        let root_dir = icu_testdata::paths::data_root()
+            .join("uprops")
```
[`uprops_toml_root()`](https://github.com/unicode-org/icu4x/blob/main/provider/testdata/src/paths.rs#L18)

echeran commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1113 @@
+            // Remove dot_above after i with upper or titlecase.
+            return Some(FullMappingResult::Remove);
+        }
+        // TODO: ICU4C has a non-standard extension for Armenian ligature ech-yiwn.
```
suggestion: go ahead

Based on the comments in [ICU-13416](https://unicode-org.atlassian.net/browse/ICU-13416), it looks like this was given a lot of thought and done at the request of a real use case with approval from various commitees' members.

I like the idea of consistency with ICU, give the work is already there (thanks!).

iainireland replied:
I'll push this off for a future patch to think about the question below.

echeran commented on experimental/casemapping/src/internals.rs:
```diff
@@ -0,0 +1,1113 @@
+        // Should we also implement it?
+        // if c == '\u{587}' {
+        //     return match (locale, is_title) {
+        // 	(CaseMapLocale::Armenian, false) => Some(FullMappingResult::String("ԵՎ")),
+        // 	(CaseMapLocale::Armenian, true) => Some(FullMappingResult::String("Եվ")),
```
question: would one of these need `CaseMapLocale::WesternArmenian`?

iainireland replied:
Good question, I'll look into it.

echeran commented on provider/uprops/src/casemapping.rs:
```diff
@@ -0,0 +1,271 @@
+}
+
+#[cfg(test)]
+mod tests {
```
suggestion: include a few more more tests

The tests `StringCaseTest::TestGreekUpper` and `StringCaseTest::TestArmenian()` seem like short tests to implement. The latter depends on whether we match ICU on Eastern vs. Western Armenian.

Also, `StringCaseTest::TestLongUpper()` seems interesting and short. Maybe Rust makes that a non-issue, but nice to have.

All the other tests seem like they are about titlecasing, UTF-8 and/or Edits, so no other tests of interest.

iainireland replied:
TestArmenian depends on enabling the Armenian code that's currently commented out.
TestGreekUpper tests a separate code path in ICU4C that I haven't ported yet.

Will make a note to revisit this in the future.

echeran commented on provider/uprops/src/uprops_serde.rs:
```diff
@@ -84,14 +84,14 @@ pub mod enumerated {
 
 pub mod script_extensions {
-    use super::enumerated::EnumeratedPropertyCodePointTrie;
+    use super::SerializedCodePointTrie;
```
praise: Thanks!

echeran reviewed (changes requested):
Looks good -- a majority of things I commented are related to info clarity in explanatory source comments.

iainireland commented on experimental/casemapping/Cargo.toml:
```diff
@@ -0,0 +1,43 @@
+std = ["icu_provider/std"]
+default = ["provider_serde"]
+provider_serde = ["serde"]
+provider_transform_internals = ["std"]
```
We already use `provider_transform_internals` in a few other places (eg the datetime component). It's documented in components/icu/src/lib.rs, under Features.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+use crate::error::Error;
+use crate::internals::{ClosureSet, DotType, MappingKind};
+
+// This represents case mapping exceptions that can't be represented as a delta applied to
```
I'll polish documentation in a future patch.

iainireland commented on experimental/casemapping/src/casemapping.rs:
```diff
@@ -0,0 +1,152 @@
+            .simple_fold(c, FoldOptions::default())
+    }
+
+    /// Returns the simple case folding mapping of the given char, using Turkic (T) mappings for
```
The case folding data in ICU defines four kinds of case folding: common (C), simple (S), full (F), and Turkic (T). 

See here: https://www.unicode.org/Public/12.1.0/ucd/CaseFolding.txt

I'll polish documentation in a future patch.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+use crate::internals::{ClosureSet, DotType, MappingKind};
+
+// This represents case mapping exceptions that can't be represented as a delta applied to
+// the original code point. Similar to ICU4C, data is stored as a u16 array. The codepoint
```
Will fix in future patch.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+impl ExceptionHeader {
+    const SLOTS_MASK: u16 = 0xff;
+
+    // Each slot is 2 u16 elements instead of 1
```
Will fix in future patch. I rewrite all this code anyway.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+    // Each slot is 2 u16 elements instead of 1
+    const DOUBLE_SLOTS_FLAG: u16 = 0x100;
+
+    const NO_SIMPLE_CASE_FOLDING_FLAG: u16 = 0x200;
```
Will fix in future patch.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+
+    // Returns whether this code point is case-sensitive.
+    // (Note that this information is stored in the trie for code points without
+    // exception data, but the exception index requires more bits than the delta.)
```
Fixed in future patch.

iainireland commented on experimental/casemapping/src/exceptions.rs:
```diff
@@ -0,0 +1,394 @@
+        let previous_slots = self.0 & previous_slot_mask;
+        let slot_num = previous_slots.count_ones() as usize;
+
+        if self.has_double_slots() {
```
This all gets rewritten in a future patch.

iainireland commented:
Okay, finally got a chance to page this back in. I've taken notes on all the feedback about improved comments/documentation, fixed up a couple of places where Elango had concrete suggestions (eg `uprops_toml_root`), and rebased. I'll address the remaining feedback in a subsequent PR. 

If I can get a stamp, I think this is ready to land.

sffc reviewed (approved):
Approval Stamp

Note that Googlers are not working tomorrow (Friday) so if you need to get this merged before Monday you may need to manually dismiss Elango's "changes requested" review.
